pkg/helm/release/manager_factory.go

```diff
@@ -63,68 +67,133 @@ func (f managerFactory) NewManager(cr *unstructured.Unstructured) (Manager, erro
-		storageBackend:   storageBackend,
-		tillerKubeClient: tillerKubeClient,
-		chartDir:         f.chartDir,
+		actionConfig:   actionConfig,
+		storageBackend: v3StorageBackend,
+		kubeClient:     ownerRefClient,
 
-		tiller:      releaseServer,
 		releaseName: releaseName,
 		namespace:   cr.GetNamespace(),
 
-		spec:   cr.Object["spec"],
+		chart:  crChart,
+		values: values,
 		status: types.StatusFor(cr),
 	}, nil
 }
 
-// getReleaseServer creates a ReleaseServer configured with a rendering engine that adds ownerrefs to rendered assets
-// based on the CR.
-func getReleaseServer(cr *unstructured.Unstructured, storageBackend *storage.Storage, tillerKubeClient *kube.Client) (*tiller.ReleaseServer, error) {
-	controllerRef := metav1.NewControllerRef(cr, cr.GroupVersionKind())
-	ownerRefs := []metav1.OwnerReference{
-		*controllerRef,
+type ownerRefInjectingClient struct {
+	refs []metav1.OwnerReference
+	kube.Client
+}
+
+func (c *ownerRefInjectingClient) Build(reader io.Reader, validate bool) (kube.ResourceList, error) {
+	resourceList, err := c.Client.Build(reader, validate)
+	if err != nil {
+		return resourceList, err
 	}
-	baseEngine := helmengine.New()
-	restMapper, err := tillerKubeClient.Factory.ToRESTMapper()
+	err = resourceList.Visit(func(r *resource.Info, err error) error {
+		if err != nil {
+			return err
+		}
+		objMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(r.Object)
+		if err != nil {
+			return err
+		}
+		u := &unstructured.Unstructured{Object: objMap}
+		if r.ResourceMapping().Scope == meta.RESTScopeNamespace {
+			u.SetOwnerReferences(c.refs)
+		}
+		return nil
+	})
 	if err != nil {
 		return nil, err
 	}
-	e := engine.NewOwnerRefEngine(baseEngine, restMapper, ownerRefs)
-	var ey environment.EngineYard = map[string]environment.Engine{
-		environment.GoTplEngine: e,
+	return resourceList, nil
+}
+
+func convertV2ToV3(v2StorageBackend *v2storage.Storage, v3StorageBackend *v3storage.Storage, cr *unstructured.Unstructured) error {
+	// If a v2 release with the legacy name exists, convert it to v3.
+	legacyName := getLegacyName(cr)
+	legacyHistoryV2, legacyV2Exists, err := releaseHistoryV2(v2StorageBackend, legacyName)
+	if err != nil {
+		return err
 	}
-	env := &environment.Environment{
-		EngineYard: ey,
-		Releases:   storageBackend,
-		KubeClient: tillerKubeClient,
+	if legacyV2Exists {
+		if err := convertHistoryToV3(legacyHistoryV2, v2StorageBackend, v3StorageBackend); err != nil {
+			return err
+		}
+		return nil
 	}
-	kubeconfig, err := tillerKubeClient.ToRESTConfig()
+
+	// If a v2 release with the CR name exists, convert it to v3.
+	releaseName := cr.GetName()
+	historyV2, existsV2, err := releaseHistoryV2(v2StorageBackend, releaseName)
 	if err != nil {
-		return nil, err
+		return err
 	}
-	cs, err := clientset.NewForConfig(kubeconfig)
-	if err != nil {
-		return nil, err
+	if existsV2 {
+		if err := convertHistoryToV3(historyV2, v2StorageBackend, v3StorageBackend); err != nil {
```

Comment:
```suggestion
		return convertHistoryToV3(historyV2, v2StorageBackend, v3StorageBackend)
```
And modify below code.